### ISSUE TITLE
v5: typing errors in various places

### ISSUE THREAD

**Issue by romain-hasseveldt** [strapi/strapi#20735]
### Required System information

- Node.js version: 20.9.0
- NPM version:
- Strapi version: 5.0.0-rc.3
- Database: SQLite
- Operating system: macOS
- Is your project Javascript or Typescript: TypeScript

### Describe the bug

There are typing errors in various places in the application. Among others:

- On an entity returned by the Document Service API, the populated fields (via the populate option) do not appear, causing a typing error.
- The parameters of the Document Service API's middlewares are also incorrectly typed, which necessitates the use of `ts-ignore` in multiple places.

### Steps to reproduce the behavior

1. Use the Document Service API.
2. Use the populate option.
3. Observe that the populated fields do not appear, causing a typing error.
4. Check the parameters of the Document Service API's middlewares and see the typing issues.

### Expected behavior

The populated fields should appear correctly when using the populate option, without causing typing errors. The parameters of the middlewares in the Document Service API's middlewares should be correctly typed, eliminating the need for `ts-ignore`.

### Screenshots
<img width="496" alt="image" src="https://github.com/user-attachments/assets/2e4245eb-7b1f-4e40-b072-6d2dbeca99a5">

**Comment by Convly** [comment]
Hey Romain, I was able to reproduce the entity service issue (invalid fields present in the returned entity) and I've created a PR to solve this in #21110.

Regarding the middleware x type narrowing, could you test something like

```typescript
const { uid, action, params } = context;

if (uid === 'api::chat.chat' && action === 'create') {
	console.log(params.data);
}
```

And tell me if the inference works as expected like this :thinking:

**Comment by hanpaine** [comment]
Hi @romain-hasseveldt would you mind taking a look at @Convly 's comment above and feeding back? Thanks 🙏

**Comment by romain-hasseveldt** [comment]
> Hey Romain, I was able to reproduce the entity service issue (invalid fields present in the returned entity) and I've created a PR to solve this in #21110.
> 
> Regarding the middleware x type narrowing, could you test something like
> 
> ```ts
> const { uid, action, params } = context;
> 
> if (uid === 'api::chat.chat' && action === 'create') {
> 	console.log(params.data);
> }
> ```
> 
> And tell me if the inference works as expected like this 🤔

Hi @Convly,

Thank you for your response and for the https://github.com/strapi/strapi/pull/21110 that fixes a number of typing issues.

To get back to the topic: the inference is indeed working. However, I notice two typing issues that remain in the middlewares of the Document Service API:

- First, the result is not typed. This forces us to use ts-ignore when doing `const result = await next();`
- Second, the typing of `context.params.data` is problematic, especially with what can be passed in data when modifications are made from Strapi’s admin interface (example: `data: { users: { connect: [Array] }, createdBy: 1, updatedBy: 1 }`. The connect field is not taken into account.).

And, to get back to the [PR](https://github.com/strapi/strapi/pull/21110) that fixes the typing issues, good job 👏  Yet there are still issues with the return of the update function when populating relations, and using `documentId` in filters also causes typing problems.

Thanks for your attention.

**Comment by Convly** [comment]
Thanks for the reply!

I'll see what I can do for the middleware API

> yet there are still issues with the return of the update function when populating relations, and using documentId in filters also causes typing problems.

Do you have an example of a query that would fail?

**Comment by romain-hasseveldt** [comment]
> Thanks for the reply!
> 
> I'll see what I can do for the middleware API
> 
> > yet there are still issues with the return of the update function when populating relations, and using documentId in filters also causes typing problems.
> 
> Do you have an example of a query that would fail?

Hello @Convly, thank you for your response as well. 

Here are some examples that cause typing errors.

1) Trying to apply a filter on documentId for a relation in a count, findFirst, or findMany query. For example:

```
strapi
  .documents("api::comment.comment")
  .findFirst({ filters: { post: { documentId: postDocumentId } } });
```

This will cause an error on documentId, with the message: 'documentId' does not exist on type 'NestedAttributeCondition<"api::comment.comment", "post">'.

2) Trying to manage relations via a connect in an update query. For example:

```
strapi.documents("api::comment.comment").update({
  documentId,
  data: { 
    likers: {
      connect: [userDocumentId]
    }
  }
});
```

This will cause an error on post: Cannot assign type '{ connect: string[]; }' to type 'ID | undefined'.ts(2322)

3) More broadly, in create or update methods, when using filters or other features that are subject to typing errors (currently fixed with ts-ignore), the populate no longer works correctly with TypeScript. For example (ignore the actual logic of the query here):

```
const comment = await strapi.documents("api::comment.comment").create({
  data: { text: "Hello", user: "fakeUserDocumentId" },
  populate: { post: { filters: { id: "fakePostId" } } },
});
```

Here, there's no error on the filter, the populate works, and I can access comment.post.

```
const comment = await strapi.documents("api::comment.comment").create({
  data: { text: "Hello", user: "fakeUserDocumentId" },
  populate: { post: { filters: { documentId: "fakePostDocumentId" } } },
});
```
Here, there is an error on the filter: Cannot assign type '{ post: { filters: { documentId: string; }; }; }' to type 'Any<"api::comment.comment"> | undefined'.

The types of the property 'post' are incompatible. Cannot assign type '{ filters: { documentId: string; }; }' to type 'boolean | NestedParams<"api::post.post"> | undefined'.

The types of the property 'filters' are incompatible. An object literal can only specify known properties, and 'documentId' does not exist on type 'Any<"api::post.post">'.

This leads to cascading typing errors, because now comment.post is no longer accessible for TypeScript:

`comment.post
`

The property 'post' does not exist on type 'ContentType<"api::comment.comment", NonPopulatableAttributeNames<"api::comment.comment">>'.

Let me know if you need further adjustments or details!

**Comment by Convly** [comment]
Me again :wave:, I've created a PR for the issues you've mentioned: #21533 

Also, here's an experimental version with the changes if you want to try and see if it fixes your issues:
`0.0.0-experimental.2e2b694fbeb6a64dd7504ce6bccecf956e9eaf03`

**Comment by romain-hasseveldt** [comment]
Well done! All the issues from the previous comment are fixed in the experimental version (-36 `@ts-ignore`!). The typing of the middlewares is still pending, but that's another topic.

**Comment by romain-hasseveldt** [comment]
> * First, the result is not typed. This forces us to use ts-ignore when doing `const result = await next();`
> * Second, the typing of `context.params.data` is problematic, especially with what can be passed in data when modifications are made from Strapi’s admin interface (example: `data: { users: { connect: [Array] }, createdBy: 1, updatedBy: 1 }`. The connect field is not taken into account.).

https://github.com/strapi/strapi/issues/21552

**Comment by romain-hasseveldt** [comment]
> Glad to hear that!
> 
> If you don't mind, could you open a separate issue for the middleware types?
> 
> It would help us track what has and hasn't been fixed and work more atomically on issues like this.
> 
> _(basically, we'll close this one once the PR is merged and we'll still be able to track the middleware issue)_

https://github.com/strapi/strapi/issues/21552
